File: daemon-mac/Sources/FocusLockDaemon/ParentAuditService.swift
import Foundation

/// Append-only audit log for parental-control events. Mirrors the sessions.jsonl
/// pattern: one JSON record per line at /Library/Application Support/FocusLock/parent.audit.jsonl.
/// Surfaced to a verified parent via get_parent_audit IPC.
final class ParentAuditService {
    private static let stateDir = URL(fileURLWithPath: "/Library/Application Support/FocusLock")
    private static let logPath = stateDir.appendingPathComponent("parent.audit.jsonl")

    private let writeLock = NSLock()
    private var permissionsApplied = false

    private let enc: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        e.dateEncodingStrategy = .iso8601
        return e
    }()
    private let dec: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    init() {
        try? FileManager.default.createDirectory(at: Self.stateDir, withIntermediateDirectories: true)
    }

    func record(_ eventType: String, command: String? = nil, detail: String? = nil) {
        let entry = ParentAuditEntry(
            timestamp: Date(),
            event: eventType,
            command: command,
            detail: detail
        )
        do {
            let data = try enc.encode(entry)
            guard var line = String(data: data, encoding: .utf8) else { return }
            line += "\n"
            writeLock.withLock {
                if let handle = FileHandle(forWritingAtPath: Self.logPath.path) {
                    handle.seekToEndOfFile()
                    handle.write(line.data(using: .utf8)!)
                    handle.closeFile()
                } else {
                    try? line.data(using: .utf8)?.write(to: Self.logPath)
                }
                ensurePermissionsLocked()
            }
        } catch {
            fputs("[parent-audit] Failed to record event \(eventType)\n", stderr)
        }
    }

    /// Returns the most recent `limit` entries, newest first.
    func recent(limit: Int) -> [ParentAuditEntry] {
        guard let content = try? String(contentsOf: Self.logPath, encoding: .utf8) else { return [] }
        return content
            .components(separatedBy: "\n")
            .filter { !$0.isEmpty }
            .compactMap { try? dec.decode(ParentAuditEntry.self, from: Data($0.utf8)) }
            .suffix(limit)
            .reversed()
    }

    private func ensurePermissionsLocked() {
        // Apply chmod 600 once per process lifetime — after the first append creates
        // the file. Mirrors parent.cred / parent.tokenkey: only root can read.
        if permissionsApplied { return }
        chmod(Self.logPath.path, 0o600)
        permissionsApplied = true
    }
}

struct ParentAuditEntry: Codable {
    var timestamp: Date
    var event: String
    var command: String?
    var detail: String?
}

/// Wire-stable event names matching the Windows daemon and shared/protocol.ts.
enum ParentAuditEvents {
    static let pinSet              = "pin_set"
    static let pinChanged          = "pin_changed"
    static let pinCleared          = "pin_cleared"
    static let pinVerifySuccess    = "pin_verify_success"
    static let pinVerifyFail       = "pin_verify_fail"
    static let pinVerifyRateLimit  = "pin_verify_rate_limited"
    static let gateBlocked         = "gate_blocked"
    static let gateAllowed         = "gate_allowed"
    static let familyPaired        = "family_paired"
    static let familyUnpaired      = "family_unpaired"
    static let familyOffline5Min   = "family_offline_5min"
    static let familyReconnected   = "family_reconnected"
    static let familyCacheTampered = "family_cache_tampered"
    static let uninstallAuthorized = "uninstall_authorized"
}

File: daemon-win/FocusLock.Daemon/Services/ParentAuditService.cs
using System.Security.AccessControl;
using System.Security.Principal;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace FocusLock.Daemon.Services;

/// <summary>
/// Append-only audit log for parental-control events. Mirrors the sessions.jsonl
/// pattern: one JSON record per line at %ProgramData%\FocusLock\parent.audit.jsonl.
/// Surfaced to a verified parent via get_parent_audit IPC.
/// </summary>
public sealed class ParentAuditService
{
    private static readonly string StateDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData),
        "FocusLock");

    private static readonly string LogPath = Path.Combine(StateDir, "parent.audit.jsonl");

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    };

    private readonly ILogger<ParentAuditService> _log;
    private readonly object _writeLock = new();
    private bool _aclApplied;

    public ParentAuditService(ILogger<ParentAuditService> log)
    {
        _log = log;
        Directory.CreateDirectory(StateDir);
    }

    public void Record(string eventType, string? command = null, string? detail = null)
    {
        var entry = new ParentAuditEntry
        {
            Timestamp = DateTime.UtcNow,
            Event = eventType,
            Command = command,
            Detail = detail,
        };
        try
        {
            var json = JsonSerializer.Serialize(entry, JsonOpts);
            lock (_writeLock)
            {
                File.AppendAllText(LogPath, json + Environment.NewLine);
                EnsureAcl();
            }
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Failed to record parent audit event {Event}", eventType);
        }
    }

    /// <summary>Returns the most recent <paramref name="limit"/> entries, newest first.</summary>
    public IReadOnlyList<ParentAuditEntry> Recent(int limit)
    {
        if (!File.Exists(LogPath)) return Array.Empty<ParentAuditEntry>();
        try
        {
            var lines = File.ReadAllLines(LogPath);
            return lines
                .Reverse()
                .Take(limit)
                .Select(l => JsonSerializer.Deserialize<ParentAuditEntry>(l, JsonOpts))
                .Where(e => e != null)
                .Cast<ParentAuditEntry>()
                .ToList();
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Failed to read parent audit log");
            return Array.Empty<ParentAuditEntry>();
        }
    }

    private void EnsureAcl()
    {
        // Apply ACL once per process lifetime — after the first append creates the file.
        // Mirrors parent.cred / parent.tokenkey: SYSTEM + Administrators only.
        if (_aclApplied) return;
        try
        {
            var info = new FileSecurity();
            info.SetAccessRuleProtection(true, false);
            info.AddAccessRule(new FileSystemAccessRule(
                new SecurityIdentifier(WellKnownSidType.LocalSystemSid, null),
                FileSystemRights.FullControl, AccessControlType.Allow));
            info.AddAccessRule(new FileSystemAccessRule(
                new SecurityIdentifier(WellKnownSidType.BuiltinAdministratorsSid, null),
                FileSystemRights.FullControl, AccessControlType.Allow));
            new FileInfo(LogPath).SetAccessControl(info);
            _aclApplied = true;
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Could not restrict ACL on parent audit log");
        }
    }
}

public sealed class ParentAuditEntry
{
    public DateTime Timestamp { get; set; }
    public string Event { get; set; } = string.Empty;
    public string? Command { get; set; }
    public string? Detail { get; set; }
}

/// <summary>Wire-stable event names for the parent audit log.</summary>
public static class ParentAuditEvents
{
    public const string PinSet              = "pin_set";
    public const string PinChanged          = "pin_changed";
    public const string PinCleared          = "pin_cleared";
    public const string PinVerifySuccess    = "pin_verify_success";
    public const string PinVerifyFail       = "pin_verify_fail";
    public const string PinVerifyRateLimit  = "pin_verify_rate_limited";
    public const string GateBlocked         = "gate_blocked";
    public const string GateAllowed         = "gate_allowed";
    public const string FamilyPaired        = "family_paired";
    public const string FamilyUnpaired      = "family_unpaired";
    public const string FamilyOffline5Min   = "family_offline_5min";
    public const string FamilyReconnected   = "family_reconnected";
    public const string FamilyCacheTampered = "family_cache_tampered";
}
